cli/src/diff_util.rs

```diff
@@ -346,19 +347,65 @@ fn show_color_words_diff_line(
     Ok(())
 }
 
-fn diff_content(path: &RepoPath, value: MaterializedTreeValue) -> Result<Vec<u8>, CommandError> {
+struct FileContent {
+    /// true if this file is likely text; false if it is definitely binary.
+    is_binary: bool,
+    // TODO: maybe this could store an `enum { String, Vec<u8> }` instead? but then we have to
+    // UTF-8 validate the whole file.
+    contents: Vec<u8>,
+}
+
+impl FileContent {
+    pub(crate) fn is_empty(&self) -> bool {
+        self.contents.is_empty()
+    }
+}
+
+impl From<String> for FileContent {
+    fn from(contents: String) -> Self {
+        FileContent {
+            is_binary: false,
+            contents: contents.into_bytes(),
+        }
+    }
+}
+
+fn diff_file_content(reader: &mut dyn io::Read) -> io::Result<FileContent> {
```

Comment:
I would call this `file_content_for_diff`, or something that doesn't make it sound like this diffs file content.